src/lsm/table_mutable.zig

```diff
@@ -111,13 +119,20 @@ pub fn TableMutableType(comptime Table: type) type {
         }
 
         pub fn remove(table: *TableMutable, value: *const Value) void {
-            // If the key is already present in the hash map, the old key will not be overwritten
-            // by the new one if using e.g. putAssumeCapacity(). Instead we must use the lower
-            // level getOrPut() API and manually overwrite the old key.
-            const upsert = table.values.getOrPutAssumeCapacity(value.*);
-            upsert.key_ptr.* = tombstone_from_key(key_from_value(value));
+            const existing = table.values.fetchRemove(value.*);
+            if (usage == .secondary_index and existing != null) {
+                // Put and remove cancel each other out.
+                assert(!tombstone(&existing.?.key));
+            } else {
+                table.values.putAssumeCapacityNoClobber(tombstone_from_key(key_from_value(value)), {});
+            }
 
             assert(table.values.count() <= table.value_count_max);
+
+            if (table.values_cache) |cache| {
+                cache.insert(key_from_value(value)).* =
+                    tombstone_from_key(key_from_value(value));
+            }
```

Comment:
> I don't understand why the bulk-update doesn't work here

The existing code already updated the cache during put but not during remove. This was not a bug because get always checks values first before the cache. But if we remove the values entry instead of replacing it with a tombstone then we get a stale read from the cache.

If the cache is only supposed to be updated in bulk, we should probably remove the update in both put and remove?

Reply:
Related - it doesn't seem that the cache knows about tombstones. Does it make sense to fill the cache with tombstones, or would it be better to treat those as a cache remove?

Reply:
I tentatively removed the cache updates in put and remove and so far it doesn't cause any fuzzing failures.